penn/studyspaces.py

```diff
@@ -9,85 +15,76 @@ class StudySpaces(object):
-        """
-        l = original.split("-")
-        final = [l[1], l[2], l[0]]
-        return '-'.join(final)
+    def get_buildings(self):
+        """Returns a list of building IDs, building names, and services."""
 
-    def get_id_json(self):
-        """Makes JSON with each element associating URL, ID, and building
-        name.
-        """
-        group_study_codes = []
-        url = BASE_URL + "/booking/vpdlc"
-        soup = BeautifulSoup(requests.get(url).text, 'html5lib')
-        l = soup.find('select', {'id': 'lid'}).find_all('option')
-        for element in l:
-            if element['value'] != '0':
-                url2 = "{}/spaces?lid={}".format(BASE_URL, str(element['value']))
-                new_dict = {}
-                new_dict['id'] = int(str(element['value']))
-                new_dict['name'] = str(element.contents[0])
-                new_dict['url'] = url2
-                group_study_codes.append(new_dict)
-        return group_study_codes
+        soup = BeautifulSoup(requests.get("{}/spaces".format(BASE_URL)).content, "html5lib")
+        options = soup.find("select", {"id": "lid"}).find_all("option")
+        return [{"id": int(opt["value"]), "name": str(opt.text), "service": "libcal"} for opt in options]
 
-    def get_id_dict(self):
-        """Extracts the ID's of the room into a dictionary. Used as a
-        helper for the extract_times method.
-        """
-        group_study_codes = {}
-        url = BASE_URL + "/booking/vpdlc"
-        soup = BeautifulSoup(requests.get(url).text, 'html5lib')
-        options = soup.find('select', {'id': 'lid'}).find_all('option')
-        for element in options:
-            if element['value'] != '0':
-                group_study_codes[int(str(element['value']))] = str(element.contents[0])
-        return group_study_codes
-
-    def extract_times(self, id, date, name):
-        """Scrapes the avaiable rooms with the given ID and date.
-
-        :param id: the ID of the building
-        :param date: the date to acquire available rooms from
-        :param name: the name of the building; obtained via get_id_dict
-        """
-        url = BASE_URL + "/rooms_acc.php?gid=%s&d=%s&cap=0" % (int(id), date)
-        soup = BeautifulSoup(requests.get(url).text, 'html5lib')
-
-        time_slots = soup.find_all('form')
-        unparsed_rooms = time_slots[1].contents[2:-2]
+    @staticmethod
+    def parse_date(date):
+        date = datetime.datetime.strptime(date, "%Y-%m-%d %H:%M:%S")
+        return pytz.timezone("US/Eastern").localize(date)
 
-        roomTimes = []
+    @staticmethod
+    def get_room_id_name_mapping(building):
+        data = requests.get("{}/spaces?lid={}".format(BASE_URL, building)).content.decode("utf8")
+        # find all of the javascript room definitions
+        out = {}
+        for item in re.findall(r"resources.push\(((?s).*?)\);", data, re.MULTILINE):
+            items = {k: v for k, v in re.findall(r'(\w+?):\s*(.*?),', item)}
```

Comment:
Could you instead use the Python JSON parsing once you've found the data by regex in the previous line?

Reply:
Tried that, the issue is that the keys are not enclosed in quotes, which Python thinks is invalid JSON syntax. I figured that parsing the values directly would be better than enclosing the keys in quotes using regex and then using Python JSON parsing.